crates/pessimistic-proof-core/src/aggchain_proof.rs

```diff
@@ -0,0 +1,103 @@
+//! Aggchain proof data structures.
+//!
+//! The aggchain-proof is the extra custom logic which is verified within the
+//! pessimistic-proof.
+//!
+//! For now, this is constraint to be either one ECDSA signature, or one SP1
+//! stark proof proving a specified statement which can be abstracted here.
+use agglayer_primitives::{Address, Signature};
+use serde::{Deserialize, Serialize};
+use sha2::{Digest as Sha256Digest, Sha256};
+use sp1_zkvm::lib::utils::words_to_bytes_le;
+
+use crate::keccak::{digest::Digest, keccak256_combine};
+
+#[derive(Debug, Deserialize, Serialize, Clone, Copy)]
+pub enum AggchainType {
+    ECDSA = 0,
+    SP1 = 1,
+}
+
+pub type Vkey = [u32; 8];
+pub type StarkVkey = Vec<u8>;
+pub type StarkProof = Vec<u8>;
+
+/// Aggchain Proof which is either one ECDSA signature, or one stark proof.
+/// Contains all the necessary data for verification.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum AggchainProofData {
+    /// ECDSA signature.
+    ECDSA(AggchainProofECDSAData),
+    /// STARK proof generated with SP1 and its metadata.
+    SP1(AggchainProofSP1Data),
+}
+
+impl AggchainProofData {
+    /// Returns the aggchain hash
+    pub fn aggchain_hash(&self) -> Digest {
+        match &self {
+            AggchainProofData::ECDSA(aggchain_proof_ecdsa) => keccak256_combine([
+                &(AggchainType::ECDSA as u32).to_be_bytes(),
+                aggchain_proof_ecdsa.signer.as_slice(),
+            ]),
+            AggchainProofData::SP1(aggchain_proof_sp1) => keccak256_combine([
+                &(AggchainType::SP1 as u32).to_be_bytes(),
+                words_to_bytes_le(&aggchain_proof_sp1.aggchain_vkey).as_slice(),
+                aggchain_proof_sp1.aggchain_params.as_slice(),
+            ]),
+        }
+    }
+}
+
+/// ECDSA variant of the aggchain proof.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct AggchainProofECDSAData {
+    /// Signer committing to the state transition.
+    pub signer: Address,
+    /// Signature committing to the state transition.
+    pub signature: Signature,
+}
+
+/// SP1 variant of the aggchain proof.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct AggchainProofSP1Data {
+    /// Chain-specific commitment forwarded by the PP.
+    pub aggchain_params: Digest,
+    /// SP1 verifying key for the SP1 aggchain proof program.
+    pub aggchain_vkey: Vkey,
+    /// STARK proof.
+    pub stark_proof: StarkProof,
```

Comment:
It seems to me that this field does could be deleted. The proof is fed in separately using `stdin.write_proof(..)` and read implicitly when `verify_sp1_proof` is called. So it does not need to be included in the multi batch header.

Reply:
yeah true 👍  only needed for the structure contained in the certificate